File: kalitech-engine/src/main/resources/kalitech/engine/modules/Controllers/Controllers.js
// FILE: resources/kalitech/engine/modules/Controllers/Controllers.js
"use strict";

function req(v, msg) {
    if (v == null) throw new Error(msg);
    return v;
}

function reqStr(s, msg) {
    if (typeof s !== "string" || !s) throw new Error(msg);
    return s;
}

function loadCore() {
    const {ControllerStack} = require("./helpers/ControllerStack");
    const {ControllerRegistry} = require("./helpers/ControllerRegistry");
    const {ensureControllersHub} = require("./helpers/EngineControllers");
    const {EntityController} = require("./helpers/EntityController");
    return {ControllerStack, ControllerRegistry, ensureControllersHub, EntityController};
}

function create(engine, K) {
    req(engine, "[Controllers] engine is required");
    req(K, "[Controllers] root state is required");

    const {ControllerStack, ControllerRegistry, ensureControllersHub, EntityController} = loadCore();
    const hub = ensureControllersHub(engine, K);

    function resolveRegistry(nameOrRegistry) {
        if (typeof nameOrRegistry === "string") {
            const name = reqStr(nameOrRegistry, "[Controllers] registry name is required");

            let r = (hub && typeof hub.get === "function") ? hub.get(name) : null;
            if (!r) {
                r = new ControllerRegistry(name);
                if (hub && typeof hub.set === "function") hub.set(r);
            }
            return r;
        }
        return req(nameOrRegistry, "[Controllers] registry is required");
    }

    function buildStack(registryOrName, ctx, entity, cfg) {
        const registry = resolveRegistry(registryOrName);
        ctx = req(ctx, "[Controllers] ctx is required");
        entity = req(entity, "[Controllers] entity is required");

        if (typeof registry.build !== "function") {
            throw new Error("[Controllers] registry.build(ctx,entity,cfg) is required");
        }

        const built = registry.build(ctx, entity, cfg || null);
        const stack = new ControllerStack(built.list, built.ids);
        if (typeof stack.bind === "function") stack.bind(ctx, entity);
        return stack;
    }

    function absorbRegistratorExport(exp, id) {
        if (typeof exp === "function") {
            exp(hub, engine, K);
            return;
        }

        if (exp && Array.isArray(exp.registries)) {
            for (let i = 0; i < exp.registries.length; i++) hub.set(exp.registries[i]);
            return;
        }

        let ok = false;
        if (exp && typeof exp === "object") {
            for (const k of Object.keys(exp)) {
                if (k.indexOf("create") !== 0 || k.lastIndexOf("Registry") !== k.length - "Registry".length) continue;
                const fn = exp[k];
                if (typeof fn !== "function") continue;
                const r = fn();
                hub.set(r);
                ok = true;
            }
        }

        if (!ok) throw new Error("[Controllers] invalid registrator export: " + reqStr(id, "id"));
    }

    // self-install hot reload hook ONCE (no bootstrap edits required)
    function tryInstallHotReload(apiObj) {
        try {
            // 1) find a SystemContext-like object
            // Prefer K.ctx / K.context, otherwise any object attached by engine scripts.
            const ctx =
                (K && (K.ctx || K.context)) ||
                (engine && (engine.ctx || engine.context)) ||
                null;

            if (!ctx) return false;

            // 2) locate hotReload domain: ctx.hotReload() OR ctx.hotReload
            let hr = null;
            try {
                if (typeof ctx.hotReload === "function") hr = ctx.hotReload();
                else hr = ctx.hotReload;
            } catch (_) {
                hr = null;
            }
            if (!hr || typeof hr.register !== "function") return false;

            // 3) idempotent guard: store flag in ctx.stateDomain OR ctx.state OR ctx.state()
            // We never throw here: hot reload must be zero-crash.
            const FLAG = "__CTRL_HR_INSTALLED__";

            function getStateDomain() {
                try {
                    if (ctx.stateDomain && typeof ctx.stateDomain.get === "function") return ctx.stateDomain;
                } catch (_) {
                }
                try {
                    if (typeof ctx.state === "function") return ctx.state();
                } catch (_) {
                }
                try {
                    if (ctx.state && typeof ctx.state.get === "function") return ctx.state;
                } catch (_) {
                }
                return null;
            }

            const sd = getStateDomain();
            if (sd && typeof sd.get === "function" && sd.get(FLAG) === true) return true;

            hr.register(() => {
                try {
                    apiObj.resetAll();
                } catch (_) {
                }
            });

            if (sd && typeof sd.set === "function") sd.set(FLAG, true);

            return true;
        } catch (_) {
            return false;
        }
    }

    const api = Object.freeze({
        Registry: ControllerRegistry,

        loadRegistrators(list) {
            if (!Array.isArray(list)) return false;
            for (let i = 0; i < list.length; i++) {
                const id = String(list[i] || "");
                if (!id) continue;
                const exp = require(id);
                absorbRegistratorExport(exp, id);
            }
            return true;
        },

        registry(name) {
            return resolveRegistry(reqStr(name, "[Controllers] registry name is required"));
        },

        registerRegistry(registry) {
            registry = req(registry, "[Controllers] registry is required");
            if (!hub || typeof hub.set !== "function") {
                throw new Error("[Controllers] controllers hub has no set(registry)");
            }
            hub.set(registry);
            return registry;
        },

        register(name, fn) {
            name = reqStr(name, "[Controllers] name is required");
            fn = req(fn, "[Controllers] fn is required");
            if (typeof fn !== "function") throw new Error("[Controllers] register(name, fn): fn must be function");
            const r = resolveRegistry(name);
            fn(r);
            if (hub && typeof hub.set === "function") hub.set(r);
            return r;
        },

        // FULL WIPE one registry
        reset(name) {
            name = reqStr(name, "[Controllers] name is required");
            if (hub && typeof hub.reset === "function") return hub.reset(name);
            // fallback
            const r = resolveRegistry(name);
            if (r && typeof r.clear === "function") r.clear();
            return r;
        },

        // FULL WIPE all registries
        resetAll() {
            if (hub && typeof hub.resetAll === "function") hub.resetAll();
            return true;
        },

        stack(registryOrName, ctx, entity, cfg) {
            return buildStack(registryOrName, ctx, entity, cfg);
        },

        entity(registryOrName, ctx, entity, cfg) {
            const stack = buildStack(registryOrName, ctx, entity, cfg);
            return new EntityController(ctx, entity, stack);
        }
    });

    // install hook best-effort
    tryInstallHotReload(api);

    return api;
}

create.META = {
    moduleId: "controllers",
    globalName: "CTRL",
    version: "1.0.5",
    description: "ENGINE.controllers hub space + game-side registration + reset/resetAll + self-installed hot reload hook.",
    engineMin: "0.1.0"
};

module.exports = create;

File: kalitech-engine/src/main/resources/kalitech/engine/modules/Controllers/Controllers.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local Arrays = luaRuntime.array
local Tables = luaRuntime.table
local Classes = luaRuntime.class
local Error = luaRuntime.Error
function req(self, v, msg)
    if v == nil then
        error(
            Classes:construct(Error, msg),
            0
        )
    end
    return v
end
function reqStr(self, s, msg)
    if KTypeOf(s) ~= "string" or not s then
        error(
            Classes:construct(Error, msg),
            0
        )
    end
    return s
end
function loadCore(self)
    local lua_require_result_0 = require("./helpers/ControllerStack")
    local ControllerStack = lua_require_result_0.ControllerStack
    local lua_require_result_1 = require("./helpers/ControllerRegistry")
    local ControllerRegistry = lua_require_result_1.ControllerRegistry
    local lua_require_result_2 = require("./helpers/EngineControllers")
    local ensureControllersHub = lua_require_result_2.ensureControllersHub
    local lua_require_result_3 = require("./helpers/EntityController")
    local EntityController = lua_require_result_3.EntityController
    return {ControllerStack = ControllerStack, ControllerRegistry = ControllerRegistry, ensureControllersHub = ensureControllersHub, EntityController = EntityController}
end
create = setmetatable(
    {},
    {__call = function(lua_, self, engine, K)
        req(_G, engine, "[Controllers] engine is required")
        req(_G, K, "[Controllers] root state is required")
        local lua_loadCore_result_4 = loadCore(_G)
        local ControllerStack = lua_loadCore_result_4.ControllerStack
        local ControllerRegistry = lua_loadCore_result_4.ControllerRegistry
        local ensureControllersHub = lua_loadCore_result_4.ensureControllersHub
        local EntityController = lua_loadCore_result_4.EntityController
        local hub = ensureControllersHub(_G, engine, K)
        local function resolveRegistry(self, nameOrRegistry)
            if KTypeOf(nameOrRegistry) == "string" then
                local name = reqStr(_G, nameOrRegistry, "[Controllers] registry name is required")
                local lua_temp_5
                if hub and KTypeOf(hub.get) == "function" then
                    lua_temp_5 = hub:get(name)
                else
                    lua_temp_5 = nil
                end
                local r = lua_temp_5
                if not r then
                    r = Classes:construct(ControllerRegistry, name)
                    if hub and KTypeOf(hub.set) == "function" then
                        hub:set(r)
                    end
                end
                return r
            end
            return req(_G, nameOrRegistry, "[Controllers] registry is required")
        end
        local function buildStack(self, registryOrName, ctx, entity, cfg)
            local registry = resolveRegistry(_G, registryOrName)
            ctx = req(_G, ctx, "[Controllers] ctx is required")
            entity = req(_G, entity, "[Controllers] entity is required")
            if KTypeOf(registry.build) ~= "function" then
                error(
                    Classes:construct(Error, "[Controllers] registry.build(ctx,entity,cfg) is required"),
                    0
                )
            end
            local built = registry:build(ctx, entity, cfg or nil)
            local stack = Classes:construct(ControllerStack, built.list, built.ids)
            if KTypeOf(stack.bind) == "function" then
                stack:bind(ctx, entity)
            end
            return stack
        end
        local function absorbRegistratorExport(self, exp, id)
            if KTypeOf(exp) == "function" then
                exp(_G, hub, engine, K)
                return
            end
            if exp and Arrays:isArray(exp.registries) then
                do
                    local i = 0
                    while i < KLength(exp.registries) do
                        hub:set(KIndex(exp.registries, i))
                        i = i + 1
                    end
                end
                return
            end
            local ok = false
            if exp and KTypeOf(exp) == "table" then
                for lua_, k in ipairs(Tables:keys(exp)) do
                    do
                        if (string.find(k, "create", nil, true) or 0) - 1 ~= 0 or KString:lastIndexOf(k, "Registry") ~= #k - #"Registry" then
                            goto lua_continue21
                        end
                        local fn = exp[k]
                        if KTypeOf(fn) ~= "function" then
                            goto lua_continue21
                        end
                        local r = fn(_G)
                        hub:set(r)
                        ok = true
                    end
                    ::lua_continue21::
                end
            end
            if not ok then
                error(
                    Classes:construct(
                        Error,
                        "[Controllers] invalid registrator export: " .. reqStr(_G, id, "id")
                    ),
                    0
                )
            end
        end
        local function tryInstallHotReload(self, apiObj)
            do
                local function lua_catch(_)
                    return true, false
                end
                local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
                    local ctx = K and (K.ctx or K.context) or engine and (engine.ctx or engine.context) or nil
                    if not ctx then
                        return true, false
                    end
                    local hr = nil
                    do
                        local function lua_catch(_)
                            hr = nil
                        end
                        local lua_try, lua_hasReturned = pcall(function()
                            if KTypeOf(ctx.hotReload) == "function" then
                                hr = ctx:hotReload()
                            else
                                hr = ctx.hotReload
                            end
                        end)
                        if not lua_try then
                            lua_catch(lua_hasReturned)
                        end
                    end
                    if not hr or KTypeOf(hr.register) ~= "function" then
                        return true, false
                    end
                    local FLAG = "__CTRL_HR_INSTALLED__"
                    local function getStateDomain(self)
                        do
                            local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
                                if ctx.stateDomain and KTypeOf(ctx.stateDomain.get) == "function" then
                                    return true, ctx.stateDomain
                                end
                            end)
                            if lua_try and lua_hasReturned then
                                return lua_returnValue
                            end
                        end
                        do
                            local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
                                if KTypeOf(ctx.state) == "function" then
                                    return true, ctx:state()
                                end
                            end)
                            if lua_try and lua_hasReturned then
                                return lua_returnValue
                            end
                        end
                        do
                            local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
                                if ctx.state and KTypeOf(ctx.state.get) == "function" then
                                    return true, ctx.state
                                end
                            end)
                            if lua_try and lua_hasReturned then
                                return lua_returnValue
                            end
                        end
                        return nil
                    end
                    local sd = getStateDomain(_G)
                    if sd and KTypeOf(sd.get) == "function" and sd:get(FLAG) == true then
                        return true, true
                    end
                    hr:register(function()
                        do
                            pcall(function()
                                apiObj:resetAll()
                            end)
                        end
                    end)
                    if sd and KTypeOf(sd.set) == "function" then
                        sd:set(FLAG, true)
                    end
                    return true, true
                end)
                if not lua_try then
                    lua_hasReturned, lua_returnValue = lua_catch(lua_hasReturned)
                end
                if lua_hasReturned then
                    return lua_returnValue
                end
            end
        end
        local api = KObject:freeze({
            Registry = ControllerRegistry,
            loadRegistrators = function(self, list)
                if not Arrays:isArray(list) then
                    return false
                end
                do
                    local i = 0
                    while i < #list do
                        do
                            local id = tostring(list[i + 1] or "")
                            if not id then
                                goto lua_continue49
                            end
                            local exp = require(id)
                            absorbRegistratorExport(_G, exp, id)
                        end
                        ::lua_continue49::
                        i = i + 1
                    end
                end
                return true
            end,
            registry = function(self, name)
                return resolveRegistry(
                    _G,
                    reqStr(_G, name, "[Controllers] registry name is required")
                )
            end,
            registerRegistry = function(self, registry)
                registry = req(_G, registry, "[Controllers] registry is required")
                if not hub or KTypeOf(hub.set) ~= "function" then
                    error(
                        Classes:construct(Error, "[Controllers] controllers hub has no set(registry)"),
                        0
                    )
                end
                hub:set(registry)
                return registry
            end,
            register = function(self, name, fn)
                name = reqStr(_G, name, "[Controllers] name is required")
                fn = req(_G, fn, "[Controllers] fn is required")
                if KTypeOf(fn) ~= "function" then
                    error(
                        Classes:construct(Error, "[Controllers] register(name, fn): fn must be function"),
                        0
                    )
                end
                local r = resolveRegistry(_G, name)
                fn(_G, r)
                if hub and KTypeOf(hub.set) == "function" then
                    hub:set(r)
                end
                return r
            end,
            reset = function(self, name)
                name = reqStr(_G, name, "[Controllers] name is required")
                if hub and KTypeOf(hub.reset) == "function" then
                    return hub:reset(name)
                end
                local r = resolveRegistry(_G, name)
                if r and KTypeOf(r.clear) == "function" then
                    r:clear()
                end
                return r
            end,
            resetAll = function(self)
                if hub and KTypeOf(hub.resetAll) == "function" then
                    hub:resetAll()
                end
                return true
            end,
            stack = function(self, registryOrName, ctx, entity, cfg)
                return buildStack(
                    _G,
                    registryOrName,
                    ctx,
                    entity,
                    cfg
                )
            end,
            entity = function(self, registryOrName, ctx, entity, cfg)
                local stack = buildStack(
                    _G,
                    registryOrName,
                    ctx,
                    entity,
                    cfg
                )
                return Classes:construct(EntityController, ctx, entity, stack)
            end
        })
        tryInstallHotReload(_G, api)
        return api
    end}
)
create.META = {
    moduleId = "controllers",
    version = "1.0.5",
    description = "ENGINE.controllers hub space + game-side registration + reset/resetAll + self-installed hot reload hook.",
    engineMin = "0.1.0"
}
M = create

return M
